### PR TITLE
feat(vscode): configure remote gRPC daemon address

### PR DESCRIPTION
**Pull request by cidrblock**
## Summary
- Add VS Code settings so the Abbenay extension can attach to a remote/container daemon over **gRPC** (`host:port`, typically `:50051`) with TLS, CA, and consumer token support
- When `abbenay.daemonAddress` is unset, keep existing local IPC detection + SEA/PATH auto-start
- Reject HTTP dashboard port `:8787` (IPv4 and IPv6), URL userinfo, and strip accidental schemes (`http(s)`, `grpc`, `tcp`)
- Coalesce reconnects on settings/token changes; keep connection mode from settings after a failed remote attach
- Bump `js-yaml` to `4.3.1` (GHSA-5p4m-2wfm-xmqj)

## Changes
- New settings: `abbenay.daemonAddress`, `daemonTls`, `daemonCaPath`, `daemonSslTargetName`, `daemonTokenEnv`
- Commands: **Abbenay: Set/Clear Daemon Token** (SecretStorage → `x-abbenay-token`)
- `DaemonClient` remote TCP+TLS path; channel cleanup on failed connect; sticky-address fix for remote→local
- Open Dashboard uses the remote host on `:8787` when a remote gRPC address is configured
- README / CHANGELOG + unit tests for connection-config

## Test plan
- [x] `cd packages/vscode && npm run lint`
- [x] `cd packages/vscode && npm test` (41 passing)
- [ ] Manual: local mode (empty `daemonAddress`) still auto-starts / attaches to local daemon
- [ ] Manual: container with `-p 50051:50051` + `--grpc-tls`, set address/CA/token, verify status shows `remote:… (tls)` and chat works
- [ ] Manual: clear `daemonAddress` and confirm reconnect returns to local IPC

### CODE REVIEW

File: packages/vscode/src/daemon/connection-config.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  if (value.startsWith('[')) {
+    const end = value.indexOf(']:');
+    if (end < 0) {
+      return undefined;
+    }
+    const port = Number(value.slice(end + 2));
+    if (!Number.isFinite(port) || port <= 0 || port > 65535) {
+      return undefined;
+    }
+    return value;
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/vscode/src/daemon/connection-config.ts:78]: Shared port validation via isValidGrpcPort() so the IPv6 bracket path rejects :8787 the same as IPv4/hostname. Fixed in c019ad6.

File: packages/vscode/src/test/daemon/connection-config.test.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    assert.strictEqual(normalizeDaemonAddress('http://127.0.0.1:50051'), '127.0.0.1:50051');
+    assert.strictEqual(normalizeDaemonAddress('https://host:50051/path'), 'host:50051');
+    assert.strictEqual(normalizeDaemonAddress('[::1]:50051'), '[::1]:50051');
+    assert.strictEqual(normalizeDaemonAddress(''), undefined);
+    assert.strictEqual(normalizeDaemonAddress('   '), undefined);
+    assert.strictEqual(normalizeDaemonAddress('unix:///tmp/daemon.sock'), undefined);
+    assert.strictEqual(normalizeDaemonAddress('hostname'), undefined);
+    assert.strictEqual(normalizeDaemonAddress('https://user:secret@host:50051'), undefined);
+    assert.strictEqual(normalizeDaemonAddress('http://127.0.0.1:8787'), undefined);
+  });
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/vscode/src/test/daemon/connection-config.test.ts:43]: The existing `https://user:secret@host:50051` case already includes `@` (userinfo is stripped after the scheme). Added an explicit bare `user:secret@host:50051` case plus `[::1]:8787` / `127.0.0.1:8787` coverage so both footguns are regression-tested. Fixed in c019ad6.

**Comment by cidrblock** [comment]
Addressed Copilot’s suppressed follow-ups from the post-`c019ad6` re-review in **bde73cb**:

1. **Failed remote connect mode** — `connectionMode` now stays derived from settings (no longer forced to `local` in the connect catch).
2. **Scheme stripping** — `normalizeDaemonAddress` strips any RFC3986-style scheme (`grpc://`, `tcp://`, `http(s)://`, …).
3. **Tests** — added `grpc://` / `tcp://` coverage alongside the earlier `[::1]:8787` / userinfo cases.

**Comment by cidrblock** [comment]
Follow-up to Copilot’s suppressed note on `initializeDaemon` cleanup: init failure/timeout now calls `disposeTransport()` instead of `close()`, so connection mode from settings is preserved. Fixed in 3868d83.

File: packages/vscode/src/daemon/client.ts [hunk 3]
```diff
@@ -162,61 +199,168 @@ export class DaemonClient {
...
-    this.client = createClient(proto.AbbenayDefinition, this.channel);
+    getLogger().info('[Daemon] Attaching x-abbenay-token metadata on gRPC calls');
+    return createClientFactory()
+      .use(async function* tokenMiddleware(call, options) {
+        return yield* call.next(call.request, {
+          ...options,
+          metadata: Metadata(options.metadata).set('x-abbenay-token', token),
+        });
+      })
+      .create(proto.AbbenayDefinition, channel);
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/vscode/src/daemon/client.ts:363]: False positive for nice-grpc: `Metadata.set()` returns the Metadata instance (see `nice-grpc-common` `Metadata.js` — `return metadata` after set). The chained form matches the [nice-grpc docs](https://github.com/deeplay-io/nice-grpc#default-call-options). Copilot may be confusing this with `@grpc/grpc-js` Metadata, whose `set()` returns void. Leaving unresolved for human confirmation.

**Review by sudhirverma — APPROVED** [review]
Verified: lint clean, `tsc` compile clean, unit-test logic reviewed in depth (couldn't run full `vscode-test` e2e locally — no Electron in this sandbox).

Solid PR — clean separation in `connection-config.ts` (well-tested edge cases: :8787 rejection for IPv4/IPv6, userinfo stripping, scheme stripping, token precedence), good security instincts (token never logged, SecretStorage preferred), and the connect/reconnect state machine in `client.ts`/`extension.ts` correctly handles failed-attach and rapid setting-change cases (matches the 3 follow-up hardening commits in this PR's history).

Non-blocking nitpicks for a future pass:
1. `normalizeDaemonAddress`'s explicit `unix:` prefix check is effectively dead for `unix:///path` — the URL-scheme-strip regex consumes `unix://` first; it still ends up rejected via slash-truncation, just not through the path the comment implies.
2. TLS without `daemonCaPath` silently falls back to the system trust store (warning-only log) — a self-signed container cert will fail with a possibly-unclear gRPC error rather than an upfront actionable message.
3. `setDaemonToken`/`clearDaemonToken` trigger a reconnect even in local mode, where the setting is unused.
4. `openDashboard`: `fallbackUrl` derived from `daemonAddress` always wins over the daemon's own `startWebServer` response URL when remote — fine for the common case, could diverge if the daemon binds a different interface.

LGTM to merge.

**Review by Hrithik-Gavankar — APPROVED** [review]
Changes LGTM!
